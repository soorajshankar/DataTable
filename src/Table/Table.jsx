import React, { useState, useEffect } from "react";

export const Li = ({ i, ri, columns }) => {
  columns = [{ label: "_checkbox", width: "10px" }, ...columns];
  return (
    <tr className="table-row">
      {columns.map((c, ci) => {
        const renderer =
          c.label === "_checkbox" ? <input type="checkbox" /> : i[c.id]; // (React.ReactNode | string | number),

        return (
          <td
            key={`${ci}-${ri}`}
            className="c-child child "
            {...{
              ...(c.numeric && { align: "right" }),
              ...(c.width && { width: c.width })
            }}
          >
            {renderer}
          </td>
        );
      })}
    </tr>
  );
};
export const renderColumns = columns => {
  columns = [{ label: "_checkbox" }, ...columns];
  return columns.map(({ label = "", numeric = false, width }, ci) => {
    const renderer = label === "_checkbox" ? <input type="checkbox" /> : label;
    return (
      <td
        key={ci}
        className="c-child child "
        {...{ ...(numeric && { align: "right" }), ...(width && { width }) }}
      >
        {renderer}
      </td>
    );
  });
};
export const renderColGroups = columns => {
  columns = [{ label: "_checkbox" }, ...columns];
  return (
    <colgroup>
      {columns.map(({ label = "", numeric = false, width }, ci) => {
        const renderer =
          label === "_checkbox" ? <input type="checkbox" /> : label;
        return <col></col>;
      })}
    </colgroup>
  );
};

export const getArray = ({ size = 10, current = [], data }) => {
  // if last elements are same do nothing
  if (data[data.length - 1] === current[current.length - 1]) return current;

  // else proceed
  let newState = [...current];
  const max = current[current.length - 1]
    ? current[current.length - 1]._id || 0
    : 0;
  //   debugger;
  if (max === 0) {
    // for the first item
    newState = [{ _id: 0, ...data[0] }, ...newState];
  }
  for (let i = max; i < max + size; i++) {
    // sliding window
    if (data[i + 1]) newState.push({ _id: i + 1, ...data[i + 1] });
  }
  if (newState.length > 3 * size) {
    // cleaning the sliding window for performance adding max buffer of 3xSize
    // virtualization :)
    newState.splice(0, newState.length - 2 * size);
  }
  return newState;
};

export const getPrevArray = ({ size = 10, current = [], data }) => {
  // if first elements are same do nothing
  if (data[0] === current[0]) return current;

  // else proceed
  let newState = [];
  const min = current[0]._id > 10 ? current[0]._id - 10 : 0;
  for (let i = current[0]._id; i > min; i--) {
    // sliding window
    if (data[i - 1]) newState.unshift({ _id: i - 1, ...data[i - 1] });
  }
  newState = [...newState, ...current];
  if (newState.length > 3 * size) {
    // cleaning up with window size for better performance
    newState = newState.slice(0, 2 * size);
  }
  return [...newState];
};
const Table = ({ data = [], columns = [] }) => {
  const [current, set] = useState([]);
  useEffect(() => {
    set(getArray({ size: 30, current: [], data }));
  }, []);
  const onScroll = e => {
    window.ee = { ...e };
    let o = e.target;
    if (o.scrollTop === 0) {
      // reached top most for smooth scrolling, considering this as a seperate event and handling with help of event loop
      set(current => {
        return getPrevArray({ size: 30, current, data });
      });
      o.scrollTo(1, 1);
    } else if (o.scrollTop < 150) {
      // direct window sliding, prev window
      set(getPrevArray({ size: 30, current, data }));
    } else if (o.offsetHeight + o.scrollTop >= o.scrollHeight) {
      // direct window sliding next window
      console.log("bottom");
      set(getArray({ size: 30, current, data }));
    }
  };
  return (
    <div className="App">
      <table style={{ tableLayout: "fixed" }}>
        {renderColGroups(columns)}
        <thead>
          <tr className="table-row">{renderColumns(columns)}</tr>
        </thead>
        <tbody className="c-container" onScroll={onScroll}>
          {current.map((i, ri) => (
            <Li key={i.id} {...{ i, ri, columns }} />
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default Table;
