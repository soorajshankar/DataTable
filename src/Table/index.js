import React, { useState } from "react";

const columns = [
  {
    label: "ID",
    id: "id",
    // renderer: val => <span style={{ color: "red" }}>{val}</span>,
    width: "100px",
    numeric: true
  },
  {
    label: "NAME",
    id: "name",
    numeric: false
  },
  {
    label: "IID",
    id: "id",
    numeric: true
  }
];

const Li = ({ i }) => (
  <tr className="table-row">
    {columns.map(c => {
      const renderer = i[c.id]; // (React.ReactNode | string | number),
      return (
        <td
          className="c-child child "
          {...{
            ...(c.numeric && { align: "right" }),
            ...(c.width && { width: c.width })
          }}
        >
          {renderer}
        </td>
      );
    })}
  </tr>
);
const renderColumns = columns =>
  columns.map(({ label = "", numeric = false, width }) => {
    const renderer = label;
    return (
      <td
        className="c-child child "
        {...{ ...(numeric && { align: "right" }), ...(width && { width }) }}
      >
        {renderer}
      </td>
    );
  });
const getArray = ({ size = 10, current = [], data }) => {
  // if last elements are same do nothing
  if (data[data.length - 1] === current[current.length - 1]) return current;

  // else proceed
  const newState = [...current];
  const max = current[current.length - 1] ? current[current.length - 1].id : 0;
  for (let i = max; i < max + size; i++) {
    // sliding window
    if (data[i + 1]) newState.push(data[i + 1]);
  }
  if (newState.length > 3 * size) {
    newState.splice(0, newState.length - 2 * size);
  } else if (newState.length > size) newState.splice(0, size - 1);
  return newState;
};

const getPrevArray = ({ size = 10, current = [], data }) => {
  // if first elements are same do nothing
  if (data[0].id === current[0].id) return current;

  // else proceed
  let newState = [];
  const min = current[0].id > 10 ? data[current[0].id - 10].id : 0;
  for (let i = current[0].id; i > min; i--) {
    // sliding window
    if (data[i - 1]) newState.unshift(data[i - 1]);
  }
  newState = [...newState, ...current];
  if (newState.length > 3 * size) {
    // cleaning up with window size for better performance
    newState = newState.slice(0, 2 * size);
  }
  return [...newState];
};
const Table = ({ data = [] }) => {
  const [current, set] = useState(getArray({ size: 30, current: [], data }));
  const onScroll = e => {
    window.ee = { ...e };
    let o = e.target;
    if (o.scrollTop == 0) {
      // reached top most for smooth scrolling, considering this as a seperate event and handling with help of event loop
      set(current => {
        return getPrevArray({ size: 30, current, data });
      });
      o.scrollTo(1, 1);
    } else if (o.scrollTop < 150) {
      // direct window sliding, prev window
      set(getPrevArray({ size: 30, current, data }));
    } else if (o.offsetHeight + o.scrollTop >= o.scrollHeight) {
      // direct window sliding next window
      set(getArray({ size: 30, current, data }));
    }
  };
  return (
    <div className="App">
      <table style={{ tableLayout: "fixed" }}>
        <thead>
          <tr className="table-row">{renderColumns(columns)}</tr>
        </thead>
        <tbody className="c-container" onScroll={onScroll}>
          {current.map(i => (
            <Li key={i.id} i={i} />
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default Table;
