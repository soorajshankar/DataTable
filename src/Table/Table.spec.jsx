import React from "react";
import { render } from "enzyme";
// import { render } from "@testing-library/react";

import Table from "./Table";
import getData from "../Constants/data";
it("should give test data array", () => {
  const data = getData(5000);
  console.log(data.length);
  expect(data).toHaveLength(5001);
});
it("renders without crashing", () => {
  const wrapper = render(<Table data={getData(500000)} col />);
  // console.log(wrapper.find("table"));

  expect(wrapper).toMatchSnapshot();
});
