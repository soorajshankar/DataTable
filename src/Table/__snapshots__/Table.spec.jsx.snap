// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "App",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "style": "table-layout:fixed",
        },
        "children": Array [
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [],
                "name": "col",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "colgroup",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "table-row",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "c-child child ",
                      },
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "type": "checkbox",
                          },
                          "children": Array [],
                          "name": "input",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "type": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "type": undefined,
                          },
                        },
                      ],
                      "name": "td",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                  ],
                  "name": "tr",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "thead",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "c-container",
                },
                "children": Array [],
                "name": "tbody",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "table-row",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "c-child child ",
                    },
                    "children": Array [
                      Object {
                        "attribs": Object {
                          "type": "checkbox",
                        },
                        "children": Array [],
                        "name": "input",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "type": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "type": undefined,
                        },
                      },
                    ],
                    "name": "td",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "tr",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "thead",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "c-container",
              },
              "children": Array [],
              "name": "tbody",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [],
                  "name": "col",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "colgroup",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          Object {
            "attribs": Object {
              "class": "c-container",
            },
            "children": Array [],
            "name": "tbody",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "table-row",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "c-child child ",
                      },
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "type": "checkbox",
                          },
                          "children": Array [],
                          "name": "input",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "type": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "type": undefined,
                          },
                        },
                      ],
                      "name": "td",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                  ],
                  "name": "tr",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "thead",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "col",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                ],
                "name": "colgroup",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "table",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
